src/backend/loadPlan.js

```diff
@@ -0,0 +1,50 @@
+//@flow
+
+import {Graph} from "../core/graph";
+import {type Project} from "../core/project";
+import {type TimelineCredParameters} from "../analysis/timeline/params";
+import {TimelineCred} from "../analysis/timeline/timelineCred";
+
+/**
+ * This file introduces a "plan" concept.
```

Comment:
IMO, this docstring provides the "why" (do we have a plan) but not "what" (is a plan).

I suggest something like:

---

This file introduces `Plans`. A `Plan` represents an execution context for SourceCred code that needs access to environmental variables, hidden credentials, disk, and IO with the outside world. Functionally, a `Plan` allows us to separate providing access to these environmental resources from providing parameters to control SourceCred's specific behavior.

Each `Plan` is a function which takes parameters and returns a Promise. The plan constructor is responsible for wrapping in any environmental variables; Plan consumers need only provide specific parameters and arguments.

Having Plans allows us to cleanly test SourceCred loading behavior without needing extensive dependency injection along the boundary with external data and IO...

---

Also, apropos of this discussion: why the name `Plan`? I don't find it super descriptive and wonder if we could find something more specific. Like `EnvironmentExecutor` or something? (I admit that's not great either.)

Reply:
+1 to addressing “what” first and foremost. Certainly include “why” if
that’s not obvious from the “what”.

+1 that *plan* is a strange name here. When I think of a plan, I think
of a piece of data describing how something should happen, like a query
plan (SQL) or a test plan. I don’t think of an action that executes
asynchronously and then returns nothing or something after having some
side effects.

At a glance, it looks like you’re basically trying to define monad
transformers for various `ReaderT`? …but then it’s not clear what
`MirrorPlan` is or how it fits in.

Reply:
LoadPlan would be a struct of closures, or partially applied functions.

Perhaps the components as `MirrorClosure`, `GraphClosure`, `CredComputationClosure` with `LoadPlan` make sense?

Reply:
“Closure” tells me how it’s represented but not what it is. (It’s still
better than “plan”, because “plan” purports to tell me what it is but
actually conveys a misleading message.) These look like _providers_ or
_producers_, which are also terms familiar from dependency injection
frameworks.

The distinction in form between `LoadPlan` and the other “plans” is also
confusing on a second read. Why is `LoadPlan` a product while all the
others are functions?

It is hard for me to see the bigger picture here, partly because these
PRs are split so finely that I cannot see how the pieces fit together.